File: riveredge-frontend/src/apps/kuaiplm/services/phase2.ts
/**
 * 快研发 Phase2：需求 / 设计评审 / FMEA
 */

import { apiRequest } from '../../../services/api';

const BASE = '/apps/kuaiplm/phase2';

export interface RdRequirement {
  id?: number;
  requirement_code?: string;
  title?: string;
  project_id?: number | null;
  project_name?: string | null;
  project_code?: string | null;
  source_type?: string | null;
  source_id?: number | null;
  priority?: string;
  status?: string;
  owner_name?: string | null;
  description?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface RdDesignReview {
  id?: number;
  review_code?: string;
  title?: string;
  project_id?: number | null;
  project_name?: string | null;
  project_code?: string | null;
  material_id?: number | null;
  material_code?: string | null;
  material_name?: string | null;
  review_type?: string;
  status?: string;
  review_date?: string | null;
  reviewer_name?: string | null;
  review_notes?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface RdFmeaRecord {
  id?: number;
  fmea_code?: string;
  title?: string;
  project_id?: number | null;
  project_name?: string | null;
  fmea_type?: string;
  status?: string;
  material_code?: string | null;
  material_name?: string | null;
  risk_items?: unknown[] | string | null;
  description?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

function unwrapList<T>(res: unknown): { items: T[]; total: number } {
  if (Array.isArray(res)) return { items: res as T[], total: res.length };
  const r = res as Record<string, unknown>;
  const items = (r.items ?? r.data ?? r.results ?? []) as T[];
  const total = Number(r.total ?? (Array.isArray(items) ? items.length : 0));
  return { items: Array.isArray(items) ? items : [], total };
}

export async function listRequirements(params?: {
  skip?: number;
  limit?: number;
  keyword?: string;
  project_id?: number;
  status?: string;
  priority?: string;
  requirement_code?: string;
  title?: string;
  sort_field?: string;
  sort_order?: string;
  created_start_date?: string;
  created_end_date?: string;
  updated_start_date?: string;
  updated_end_date?: string;
}) {
  const res = await apiRequest<unknown>(`${BASE}/requirements`, { method: 'GET', params });
  return unwrapList<RdRequirement>(res);
}

export async function createRequirement(data: Partial<RdRequirement>) {
  return apiRequest<RdRequirement>(`${BASE}/requirements`, { method: 'POST', data });
}

export async function updateRequirement(id: number | string, data: Partial<RdRequirement>) {
  return apiRequest<RdRequirement>(`${BASE}/requirements/${id}`, { method: 'PUT', data });
}

export async function deleteRequirement(id: number | string) {
  return apiRequest<void>(`${BASE}/requirements/${id}`, { method: 'DELETE' });
}

export async function listDesignReviews(params?: {
  skip?: number;
  limit?: number;
  keyword?: string;
  project_id?: number;
  status?: string;
  review_code?: string;
  title?: string;
  sort_field?: string;
  sort_order?: string;
  created_start_date?: string;
  created_end_date?: string;
  updated_start_date?: string;
  updated_end_date?: string;
}) {
  const res = await apiRequest<unknown>(`${BASE}/design-reviews`, { method: 'GET', params });
  return unwrapList<RdDesignReview>(res);
}

export async function createDesignReview(data: Partial<RdDesignReview>) {
  return apiRequest<RdDesignReview>(`${BASE}/design-reviews`, { method: 'POST', data });
}

export async function updateDesignReview(id: number | string, data: Partial<RdDesignReview>) {
  return apiRequest<RdDesignReview>(`${BASE}/design-reviews/${id}`, { method: 'PUT', data });
}

export async function deleteDesignReview(id: number | string) {
  return apiRequest<void>(`${BASE}/design-reviews/${id}`, { method: 'DELETE' });
}

export async function listFmeaRecords(params?: {
  skip?: number;
  limit?: number;
  keyword?: string;
  project_id?: number;
  status?: string;
  fmea_type?: string;
  fmea_code?: string;
  title?: string;
  sort_field?: string;
  sort_order?: string;
  created_start_date?: string;
  created_end_date?: string;
  updated_start_date?: string;
  updated_end_date?: string;
}) {
  const res = await apiRequest<unknown>(`${BASE}/fmea`, { method: 'GET', params });
  return unwrapList<RdFmeaRecord>(res);
}

export async function createFmeaRecord(data: Partial<RdFmeaRecord>) {
  return apiRequest<RdFmeaRecord>(`${BASE}/fmea`, { method: 'POST', data });
}

export async function updateFmeaRecord(id: number | string, data: Partial<RdFmeaRecord>) {
  return apiRequest<RdFmeaRecord>(`${BASE}/fmea/${id}`, { method: 'PUT', data });
}

export async function deleteFmeaRecord(id: number | string) {
  return apiRequest<void>(`${BASE}/fmea/${id}`, { method: 'DELETE' });
}

File: riveredge-backend/src/apps/kuaiplm/api/phase2.py
"""
二期 API：需求 / 设计评审 / FMEA

Author: RiverEdge Team
Date: 2026-05-28
"""

import uuid
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Path, Query, status
from loguru import logger

from apps.kuaiplm.schemas.phase2 import (
    RdDesignReviewCreate,
    RdDesignReviewResponse,
    RdDesignReviewUpdate,
    RdFmeaRecordCreate,
    RdFmeaRecordResponse,
    RdFmeaRecordUpdate,
    RdRequirementCreate,
    RdRequirementResponse,
    RdRequirementUpdate,
)
from apps.kuaiplm.services.phase2_service import Phase2Service
from core.api.deps.access import require_access
from core.api.deps.deps import get_current_tenant
from infra.api.deps.deps import get_current_user
from infra.exceptions.exceptions import NotFoundError
from infra.models.user import User

router = APIRouter(prefix="/phase2", tags=["App · Kuaiplm · Phase2"])
service = Phase2Service()


def _err(status_code: int, message: str, route: str) -> HTTPException:
    trace_id = uuid.uuid4().hex
    logger.warning("kuaiplm_phase2_api_error trace_id={} route={} message={}", trace_id, route, message)
    return HTTPException(status_code=status_code, detail={"message": message, "trace_id": trace_id})


# ---------- Requirements ----------

@router.get("/requirements", summary="List requirements")
async def list_requirements(
    project_id: Optional[int] = Query(None),
    status: Optional[str] = Query(None),
    keyword: Optional[str] = Query(None),
    requirement_code: Optional[str] = Query(None),
    title: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    sort_field: Optional[str] = Query(None),
    sort_order: Optional[str] = Query(None, description="asc | desc"),
    created_start_date: Optional[str] = Query(None),
    created_end_date: Optional[str] = Query(None),
    updated_start_date: Optional[str] = Query(None),
    updated_end_date: Optional[str] = Query(None),
    _auth=Depends(require_access("kuaiplm.requirement", "read", required_permissions=["kuaiplm:requirement:read"])),
    tenant_id: int = Depends(get_current_tenant),
):
    rows, total = await service.list_requirements(
        tenant_id,
        project_id=project_id,
        status=status,
        keyword=keyword,
        requirement_code=requirement_code,
        title=title,
        skip=skip,
        limit=limit,
        sort_field=sort_field,
        sort_order=sort_order,
        created_start_date=created_start_date,
        created_end_date=created_end_date,
        updated_start_date=updated_start_date,
        updated_end_date=updated_end_date,
    )
    return {"data": rows, "total": total, "success": True}


@router.post("/requirements", response_model=RdRequirementResponse, status_code=status.HTTP_201_CREATED)
async def create_requirement(
    data: RdRequirementCreate,
    current_user: User = Depends(get_current_user),
    _auth=Depends(require_access("kuaiplm.requirement", "create", required_permissions=["kuaiplm:requirement:create"])),
    tenant_id: int = Depends(get_current_tenant),
):
    return await service.create_requirement(tenant_id, data, current_user.id)


@router.put("/requirements/{requirement_id}", response_model=RdRequirementResponse)
async def update_requirement(
    data: RdRequirementUpdate,
    requirement_id: int = Path(...),
    current_user: User = Depends(get_current_user),
    _auth=Depends(require_access("kuaiplm.requirement", "update", required_permissions=["kuaiplm:requirement:update"])),
    tenant_id: int = Depends(get_current_tenant),
):
    try:
        return await service.update_requirement(tenant_id, requirement_id, data, current_user.id)
    except NotFoundError as e:
        raise _err(404, str(e), f"/phase2/requirements/{requirement_id}")


@router.delete("/requirements/{requirement_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_requirement(
    requirement_id: int = Path(...),
    current_user: User = Depends(get_current_user),
    _auth=Depends(require_access("kuaiplm.requirement", "delete", required_permissions=["kuaiplm:requirement:delete"])),
    tenant_id: int = Depends(get_current_tenant),
):
    try:
        await service.delete_requirement(tenant_id, requirement_id, current_user.id)
    except NotFoundError as e:
        raise _err(404, str(e), f"/phase2/requirements/{requirement_id}")


# ---------- Design Reviews ----------

@router.get("/design-reviews", summary="List design reviews")
async def list_design_reviews(
    project_id: Optional[int] = Query(None),
    status: Optional[str] = Query(None),
    keyword: Optional[str] = Query(None),
    review_code: Optional[str] = Query(None),
    title: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    sort_field: Optional[str] = Query(None),
    sort_order: Optional[str] = Query(None, description="asc | desc"),
    created_start_date: Optional[str] = Query(None),
    created_end_date: Optional[str] = Query(None),
    updated_start_date: Optional[str] = Query(None),
    updated_end_date: Optional[str] = Query(None),
    _auth=Depends(require_access("kuaiplm.design-review", "read", required_permissions=["kuaiplm:design-review:read"])),
    tenant_id: int = Depends(get_current_tenant),
):
    rows, total = await service.list_design_reviews(
        tenant_id,
        project_id=project_id,
        status=status,
        keyword=keyword,
        review_code=review_code,
        title=title,
        skip=skip,
        limit=limit,
        sort_field=sort_field,
        sort_order=sort_order,
        created_start_date=created_start_date,
        created_end_date=created_end_date,
        updated_start_date=updated_start_date,
        updated_end_date=updated_end_date,
    )
    return {"data": rows, "total": total, "success": True}


@router.post("/design-reviews", response_model=RdDesignReviewResponse, status_code=status.HTTP_201_CREATED)
async def create_design_review(
    data: RdDesignReviewCreate,
    current_user: User = Depends(get_current_user),
    _auth=Depends(require_access("kuaiplm.design-review", "create", required_permissions=["kuaiplm:design-review:create"])),
    tenant_id: int = Depends(get_current_tenant),
):
    return await service.create_design_review(tenant_id, data, current_user.id)


@router.put("/design-reviews/{review_id}", response_model=RdDesignReviewResponse)
async def update_design_review(
    data: RdDesignReviewUpdate,
    review_id: int = Path(...),
    current_user: User = Depends(get_current_user),
    _auth=Depends(require_access("kuaiplm.design-review", "update", required_permissions=["kuaiplm:design-review:update"])),
    tenant_id: int = Depends(get_current_tenant),
):
    try:
        return await service.update_design_review(tenant_id, review_id, data, current_user.id)
    except NotFoundError as e:
        raise _err(404, str(e), f"/phase2/design-reviews/{review_id}")


@router.delete("/design-reviews/{review_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_design_review(
    review_id: int = Path(...),
    current_user: User = Depends(get_current_user),
    _auth=Depends(require_access("kuaiplm.design-review", "delete", required_permissions=["kuaiplm:design-review:delete"])),
    tenant_id: int = Depends(get_current_tenant),
):
    try:
        await service.delete_design_review(tenant_id, review_id, current_user.id)
    except NotFoundError as e:
        raise _err(404, str(e), f"/phase2/design-reviews/{review_id}")


# ---------- FMEA ----------

@router.get("/fmea", summary="List FMEA records")
async def list_fmea_records(
    project_id: Optional[int] = Query(None),
    fmea_type: Optional[str] = Query(None),
    status: Optional[str] = Query(None),
    keyword: Optional[str] = Query(None),
    fmea_code: Optional[str] = Query(None),
    title: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    sort_field: Optional[str] = Query(None),
    sort_order: Optional[str] = Query(None, description="asc | desc"),
    created_start_date: Optional[str] = Query(None),
    created_end_date: Optional[str] = Query(None),
    updated_start_date: Optional[str] = Query(None),
    updated_end_date: Optional[str] = Query(None),
    _auth=Depends(require_access("kuaiplm.fmea", "read", required_permissions=["kuaiplm:fmea:read"])),
    tenant_id: int = Depends(get_current_tenant),
):
    rows, total = await service.list_fmea_records(
        tenant_id,
        project_id=project_id,
        fmea_type=fmea_type,
        status=status,
        keyword=keyword,
        fmea_code=fmea_code,
        title=title,
        skip=skip,
        limit=limit,
        sort_field=sort_field,
        sort_order=sort_order,
        created_start_date=created_start_date,
        created_end_date=created_end_date,
        updated_start_date=updated_start_date,
        updated_end_date=updated_end_date,
    )
    return {"data": rows, "total": total, "success": True}


@router.post("/fmea", response_model=RdFmeaRecordResponse, status_code=status.HTTP_201_CREATED)
async def create_fmea_record(
    data: RdFmeaRecordCreate,
    current_user: User = Depends(get_current_user),
    _auth=Depends(require_access("kuaiplm.fmea", "create", required_permissions=["kuaiplm:fmea:create"])),
    tenant_id: int = Depends(get_current_tenant),
):
    return await service.create_fmea_record(tenant_id, data, current_user.id)


@router.put("/fmea/{fmea_id}", response_model=RdFmeaRecordResponse)
async def update_fmea_record(
    data: RdFmeaRecordUpdate,
    fmea_id: int = Path(...),
    current_user: User = Depends(get_current_user),
    _auth=Depends(require_access("kuaiplm.fmea", "update", required_permissions=["kuaiplm:fmea:update"])),
    tenant_id: int = Depends(get_current_tenant),
):
    try:
        return await service.update_fmea_record(tenant_id, fmea_id, data, current_user.id)
    except NotFoundError as e:
        raise _err(404, str(e), f"/phase2/fmea/{fmea_id}")


@router.delete("/fmea/{fmea_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_fmea_record(
    fmea_id: int = Path(...),
    current_user: User = Depends(get_current_user),
    _auth=Depends(require_access("kuaiplm.fmea", "delete", required_permissions=["kuaiplm:fmea:delete"])),
    tenant_id: int = Depends(get_current_tenant),
):
    try:
        await service.delete_fmea_record(tenant_id, fmea_id, current_user.id)
    except NotFoundError as e:
        raise _err(404, str(e), f"/phase2/fmea/{fmea_id}")
